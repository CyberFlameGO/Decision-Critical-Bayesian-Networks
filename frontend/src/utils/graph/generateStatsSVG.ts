import { dcbn } from "./graph";

export function generateStatsSVG(
  graphValue: number[] | undefined,
  type: dcbn.NodeValueType = "empty",
  virtualEvidence: number | null = null
): string {
  const svgContainer = document.createElementNS(
    "http://www.w3.org/2000/svg",
    "svg"
  );
  svgContainer.setAttribute("xmlns", "http://www.w3.org/2000/svg");
  svgContainer.setAttribute("viewBox", "0 0 300 100");
  svgContainer.setAttribute("width", "300");
  svgContainer.setAttribute("height", "100");
  const background = document.createElementNS(
    "http://www.w3.org/2000/svg",
    "rect"
  );
  background.setAttribute("width", "100%");
  background.setAttribute("height", "100%");
  background.setAttribute(
    "fill",
    type == "evidence"
      ? "#e47833"
      : type == "empty"
      ? "#6c7a89"
      : type == "computed"
      ? "#3498db"
      : ""
  );
  background.setAttribute("rx", "4");
  background.setAttribute("ry", "4");
  svgContainer.appendChild(background);
  const line = document.createElementNS("http://www.w3.org/2000/svg", "line");
  line.setAttribute("x1", "0%");
  line.setAttribute("x2", "100%");
  line.setAttribute("y1", "50%");
  line.setAttribute("y2", "50%");
  line.setAttribute("opacity", "0.3");
  line.setAttribute("stroke", "white");
  line.setAttribute("stroke-dasharray", "17");
  svgContainer.appendChild(line);
  if (virtualEvidence !== null) {
    const virtualLine = document.createElementNS(
      "http://www.w3.org/2000/svg",
      "line"
    );
    virtualLine.setAttribute("x1", "0%");
    virtualLine.setAttribute("x2", "100%");
    virtualLine.setAttribute("y1", `${80 * (1 - virtualEvidence) + 10}%`);
    virtualLine.setAttribute("y2", `${80 * (1 - virtualEvidence) + 10}%`);
    virtualLine.setAttribute("opacity", "0.4");
    virtualLine.setAttribute("stroke", "white");
    virtualLine.setAttribute("stroke-width", "3");
    svgContainer.appendChild(virtualLine);
  }
  if (graphValue) {
    if (graphValue.length === 1) graphValue.push(graphValue[0]);

    const graphString =
      "M" +
      graphValue
        .map(
          (val, i) =>
            `${(300 / (graphValue.length - 1)) * i} ${80 * (1 - val) + 10} ${
              graphValue.length == i + 1 ? "" : "L"
            }`
        )
        .join("");
    const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
    path.setAttribute("d", graphString);
    path.setAttribute("fill", "none");
    path.setAttribute("stroke", "white");
    path.setAttribute("opacity", "1");
    path.setAttribute("stroke-width", "5");
    svgContainer.appendChild(path);
  }
  const svg = svgContainer.outerHTML;
  return "data:image/svg+xml;charset=utf-8," + encodeURIComponent(svg);
}
