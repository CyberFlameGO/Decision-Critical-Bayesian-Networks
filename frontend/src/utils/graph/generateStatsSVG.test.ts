import { generateStatsSVG } from "./generateStatsSVG";

test("Empty SVG stats", () => {
  const svg = generateStatsSVG(undefined);
  expect(svg).toMatchSnapshot();
});

test("Computed SVG stats", () => {
  const svg = generateStatsSVG([0, 0.5, 1, 0.1], "computed");
  expect(svg).toMatchSnapshot();
});

test("Virt. evidence SVG stats", () => {
  const svg = generateStatsSVG(undefined, "empty", 0.5);
  expect(svg).toMatchSnapshot();
});

test("Evidence SVG stats", () => {
  const svg = generateStatsSVG([1, 0.9, 0.8], "evidence", 0.5);
  expect(svg).toMatchSnapshot();
});
