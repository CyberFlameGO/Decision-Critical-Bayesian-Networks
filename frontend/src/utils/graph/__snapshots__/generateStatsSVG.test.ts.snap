// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Computed SVG stats 1`] = `"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20300%20100%22%20width%3D%22300%22%20height%3D%22100%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22%233498db%22%20rx%3D%224%22%20ry%3D%224%22%3E%3C%2Frect%3E%3Cline%20x1%3D%220%25%22%20x2%3D%22100%25%22%20y1%3D%2250%25%22%20y2%3D%2250%25%22%20opacity%3D%220.3%22%20stroke%3D%22white%22%20stroke-dasharray%3D%2217%22%3E%3C%2Fline%3E%3Cpath%20d%3D%22M0%2090%20L100%2050%20L200%2010%20L300%2082%20%22%20fill%3D%22none%22%20stroke%3D%22white%22%20opacity%3D%221%22%20stroke-width%3D%225%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E"`;

exports[`Empty SVG stats 1`] = `"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20300%20100%22%20width%3D%22300%22%20height%3D%22100%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22%236c7a89%22%20rx%3D%224%22%20ry%3D%224%22%3E%3C%2Frect%3E%3Cline%20x1%3D%220%25%22%20x2%3D%22100%25%22%20y1%3D%2250%25%22%20y2%3D%2250%25%22%20opacity%3D%220.3%22%20stroke%3D%22white%22%20stroke-dasharray%3D%2217%22%3E%3C%2Fline%3E%3C%2Fsvg%3E"`;

exports[`Evidence SVG stats 1`] = `"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20300%20100%22%20width%3D%22300%22%20height%3D%22100%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22%23e47833%22%20rx%3D%224%22%20ry%3D%224%22%3E%3C%2Frect%3E%3Cline%20x1%3D%220%25%22%20x2%3D%22100%25%22%20y1%3D%2250%25%22%20y2%3D%2250%25%22%20opacity%3D%220.3%22%20stroke%3D%22white%22%20stroke-dasharray%3D%2217%22%3E%3C%2Fline%3E%3Cline%20x1%3D%220%25%22%20x2%3D%22100%25%22%20y1%3D%2250%25%22%20y2%3D%2250%25%22%20opacity%3D%220.4%22%20stroke%3D%22white%22%20stroke-width%3D%223%22%3E%3C%2Fline%3E%3Cpath%20d%3D%22M0%2010%20L150%2018%20L300%2025.999999999999996%20%22%20fill%3D%22none%22%20stroke%3D%22white%22%20opacity%3D%221%22%20stroke-width%3D%225%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E"`;

exports[`Virt. evidence SVG stats 1`] = `"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20300%20100%22%20width%3D%22300%22%20height%3D%22100%22%3E%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22%236c7a89%22%20rx%3D%224%22%20ry%3D%224%22%3E%3C%2Frect%3E%3Cline%20x1%3D%220%25%22%20x2%3D%22100%25%22%20y1%3D%2250%25%22%20y2%3D%2250%25%22%20opacity%3D%220.3%22%20stroke%3D%22white%22%20stroke-dasharray%3D%2217%22%3E%3C%2Fline%3E%3Cline%20x1%3D%220%25%22%20x2%3D%22100%25%22%20y1%3D%2250%25%22%20y2%3D%2250%25%22%20opacity%3D%220.4%22%20stroke%3D%22white%22%20stroke-width%3D%223%22%3E%3C%2Fline%3E%3C%2Fsvg%3E"`;
